code/iaas/auth-logic/src/main/java/io/cattle/platform/iaas/api/request/handler/GenericWhitelistedProxy.java

```diff
@@ -0,0 +1,187 @@
+package io.cattle.platform.iaas.api.request.handler;
+
+import io.cattle.platform.archaius.util.ArchaiusUtil;
+import io.cattle.platform.iaas.api.auth.github.GithubUtils;
+import io.github.ibuildthecloud.gdapi.exception.ClientVisibleException;
+import io.github.ibuildthecloud.gdapi.request.ApiRequest;
+import io.github.ibuildthecloud.gdapi.request.handler.AbstractResponseGenerator;
+import io.github.ibuildthecloud.gdapi.util.ResponseCodes;
+
+import java.io.IOException;
+import java.io.InputStream;
+import java.io.OutputStream;
+import java.net.URI;
+import java.net.URISyntaxException;
+import java.net.URLDecoder;
+import java.util.Collections;
+import java.util.Enumeration;
+import java.util.HashSet;
+import java.util.List;
+import java.util.Set;
+
+import javax.inject.Inject;
+
+import org.apache.commons.io.IOUtils;
+import org.apache.commons.lang3.StringUtils;
+import org.apache.http.Header;
+import org.apache.http.HttpResponse;
+import org.apache.http.client.ClientProtocolException;
+import org.apache.http.client.ResponseHandler;
+import org.apache.http.client.fluent.Request;
+import org.apache.http.client.fluent.Response;
+import org.apache.http.client.utils.URIBuilder;
+
+import com.netflix.config.DynamicBooleanProperty;
+import com.netflix.config.DynamicStringProperty;
+
+public class GenericWhitelistedProxy extends AbstractResponseGenerator {
+
+    private static final DynamicBooleanProperty ALLOW_PROXY = ArchaiusUtil.getBoolean("api.proxy.allow");
+    private static final DynamicStringProperty PROXY_WHITELIST = ArchaiusUtil.getString("api.proxy.whitelist");
+    private Set<String> whitelist;
+
+    GithubUtils githubUtils;
+
+    @Override
+    protected void generate(final ApiRequest request) throws IOException {
+        if (request.getRequestVersion() == null || !ALLOW_PROXY.get())
+            return;
+
+        if (!StringUtils.equals("proxy", request.getRequestVersion())) {
+            return;
+        }
+
+        String redirect = request.getServletContext().getRequest().getRequestURI();
```

Comment:
nit: Can you just save request.getServletContext().getRequest() as a local variable since it is referred to so often